src/pkg/egress/syslog/https_batch.go

```diff
@@ -0,0 +1,116 @@
+package syslog
+
+import (
+	"bytes"
+	"crypto/tls"
+	"time"
+
+	"code.cloudfoundry.org/go-loggregator/v9/rpc/loggregator_v2"
+	metrics "code.cloudfoundry.org/go-metric-registry"
+	"code.cloudfoundry.org/loggregator-agent-release/src/pkg/egress"
+)
+
+const BATCHSIZE = 256 * 1024
+
+type HTTPSBatchWriter struct {
```

Comment:
We understand the goal of the HTTPSBatchWriter to be:
* Buffer & batch incoming envelopes.
* If the batch reaches a certain size, flush the batch to a destination.
* On some interval, flush the batch to a destination.

❓ Is that right?

Currently, `HTTPSBatchWriter` and `TriggerTimer` doesn't appear to do that. We think that code actually results in the following behaviour:
* Add one message to the batch and sleep for some interval, then flush to a destination.
* After that, add messages to the batch as they come in.
* If the batch reaches a certain size, flush the batch to a destination.

➡️ Can you please have a look at adjusting this code.

Here are some examples of ways we've done batching in the past:
* https://github.com/cloudfoundry/loggregator-agent-release/blob/main/src/pkg/otelcolclient/signal_batcher.go
* https://github.com/cloudfoundry/go-loggregator/blob/main/ingress_client.go#L484-L525

Reply:
first thing:
Yes your understanding seems right on what we try to do.

I have pushed a newer version, that should comply with the envisioned behaviour.
Tests for the new version are also added, confirming the implementation complying with the wanted behaviour.
I am currently looking through your pointers to see if I can leverage some of the implementations shown there.

Reply:
Regarding the code proposed:
The implementation for the signal batcher uses slices and an append structure.
It has information of how long the batch is going to be, speeding up the execution.
In contrast, the syslog batching feature will not know beforehand, how long batches are going to be,
Giving the edge of speed to Byte.buffers:
https://stackoverflow.com/questions/39319024/builtin-append-vs-bytes-buffer-write

Also, the actual buffering code is pretty short, and the expected type for sending data with the http client used is Byte[], so the choice still seems obvious to me.

Will follow up with a discussion on the batching structure for the ingress_client, which seems to use the same idea as what I need.

Reply:
The code regarding the triggerable timer on buffer length shown in the ingress client is from what I can tell the code I would also need for syslog batching.
I was able to get the approach I was working on into a working non blocking state, and I as a not as experienced Go programmer have a way easier time understanding what is happening with the code I provided.
However, if you just want to have a similar implementation to what is already in the codebase, I am fine with adopting it that way.

I was not able to parse out test cases that really confirm, that the implementation of the ingress_client is doing what is expected, so it was a little hard to read.